File: fixtures/python_demo/irisflow/transform.py
"""The data transformation steps."""
from __future__ import annotations

import re

from . import utils as u  # aliased module import — a resolution edge case
from .clean import NUMERIC, clean_dataset
from .utils import log_step, require_columns

SPECIES_TAG = re.compile(r"^(?P<genus>[a-z]+)?_?(?P<name>setosa|versicolor|virginica)$")


def add_ratios(df):
    """STEP 1 — derive shape features from the raw measurements."""
    df = df.copy()
    df["sepal_ratio"] = df["sepal_length"] / df["sepal_width"]
    df["petal_ratio"] = df["petal_length"] / df["petal_width"]
    return log_step("add_ratios", df)


def trim_outliers(df, max_rounds=5, z_thresh=2.0):
    """STEP 2 — iteratively trim outliers, one round at a time.

    THE LOOP FIXTURE. Nested: an outer convergence loop over rounds, an inner
    loop over species. Every iteration carries state worth stepping through
    (round, species, max z-score, rows dropped, rows remaining) and the loop
    terminates on a data-dependent condition rather than a fixed count.
    """
    history = []

    for round_idx in range(1, max_rounds + 1):
        dropped_this_round = 0

        for sp in sorted(df["species"].unique()):
            rows = df[df["species"] == sp]
            z = ((rows["petal_ratio"] - rows["petal_ratio"].mean()) / rows["petal_ratio"].std()).abs()

            if len(z) and z.max() > z_thresh:
                drop_id = z.idxmax()
                df = df.drop(index=drop_id)
                dropped_this_round += 1

            history.append(
                {
                    "round": round_idx,
                    "species": sp,
                    "max_z": float(z.max()) if len(z) else None,
                    "dropped": dropped_this_round,
                    "remaining": len(df),
                }
            )

        if dropped_this_round == 0:
            break

    df = df.reset_index(drop=True)
    df.attrs["trim_history"] = history
    return log_step("trim_outliers", df)


def scale_numeric(df, columns=None):
    """STEP 3 — z-score the measurement columns."""
    columns = columns or NUMERIC
    require_columns(df, columns)
    df = df.copy()
    for col in columns:
        df[col] = (df[col] - df[col].mean()) / df[col].std()
    return u.log_step("scale_numeric", df)  # called through the module alias


def aggregate_by_species(df):
    """STEP 4 — collapse to one row per species."""
    grouped = df.groupby("species").mean(numeric_only=True).reset_index()
    return log_step("aggregate_by_species", grouped)


def parse_species_tag(value):
    """A regex applied per-row — the fixture for the stretch-goal regex inspector.

    Every call has a clean in/out pair worth capturing: the input string, the
    pattern, whether it matched, and the named groups it produced.
    """
    match = SPECIES_TAG.match(value.strip().lower())
    if not match:
        return {"genus": None, "name": value, "matched": False}
    return {**match.groupdict(), "matched": True}


def tag_species(df):
    """Apply the regex across the frame."""
    df = df.copy()
    parsed = df["species"].map(parse_species_tag)
    df["species_matched"] = [p["matched"] for p in parsed]
    return log_step("tag_species", df)


def apply_steps(df, steps):
    """Higher-order: functions passed as values — the hard case for static resolution."""
    for fn in steps:
        df = fn(df)
    return df


def default_steps():
    return [clean_dataset, add_ratios, trim_outliers, scale_numeric, aggregate_by_species]

File: fixtures/r_demo/R/transform.R
# The data transformation steps.

suppressPackageStartupMessages({
  library(dplyr)
  library(purrr)
})

#' STEP 1 — derive shape features from the raw measurements.
add_ratios <- function(df) {
  df <- df %>%
    mutate(
      sepal_ratio = sepal_length / sepal_width,
      petal_ratio = petal_length / petal_width
    )
  log_step("add_ratios", df)
}

#' STEP 2 — iteratively trim outliers, one round at a time.
#'
#' THE LOOP FIXTURE. Nested: an outer convergence loop over rounds, an inner
#' loop over species. Every iteration carries state worth stepping through
#' (round, species, max z-score, rows dropped, rows remaining) and the loop
#' terminates on a data-dependent condition rather than a fixed count.
trim_outliers <- function(df, max_rounds = 5, z_thresh = 2.0) {
  history <- list()

  for (round_idx in seq_len(max_rounds)) {
    dropped_this_round <- 0

    for (sp in sort(unique(df$species))) {
      rows <- df %>% filter(species == sp)
      z <- abs(scale(rows$petal_ratio)[, 1])
      worst <- which.max(z)

      if (length(z) > 0 && !is.na(z[worst]) && z[worst] > z_thresh) {
        drop_id <- rows$row_id[worst]
        df <- df %>% filter(row_id != drop_id)
        dropped_this_round <- dropped_this_round + 1
      }

      history[[length(history) + 1]] <- list(
        round = round_idx,
        species = sp,
        max_z = if (length(z) > 0) max(z, na.rm = TRUE) else NA_real_,
        dropped = dropped_this_round,
        remaining = nrow(df)
      )
    }

    if (dropped_this_round == 0) break
  }

  attr(df, "trim_history") <- history
  log_step("trim_outliers", df)
}

#' STEP 3 — z-score the measurement columns.
scale_numeric <- function(df, columns = NULL) {
  if (is.null(columns)) columns <- NUMERIC
  require_columns(df, columns)
  df <- df %>% mutate(across(all_of(columns), ~ as.numeric(scale(.x))))
  log_step("scale_numeric", df)
}

#' STEP 4 — collapse to one row per species.
aggregate_by_species <- function(df) {
  grouped <- df %>%
    group_by(species) %>%
    summarise(across(where(is.numeric), mean), .groups = "drop") %>%
    select(-any_of("row_id"))
  log_step("aggregate_by_species", grouped)
}

#' Higher-order: functions passed as values — the hard case for static resolution.
apply_steps <- function(df, steps) {
  reduce(steps, function(acc, fn) fn(acc), .init = df)
}

default_steps <- function() {
  list(clean_dataset, add_ratios, trim_outliers, scale_numeric, aggregate_by_species)
}

#' A regex applied per-row — the fixture for the stretch-goal regex inspector.
#' Every call has a clean in/out pair worth capturing: the input string, the
#' pattern, whether it matched, and the groups it produced.
SPECIES_TAG <- "^([a-z]+)?_?(setosa|versicolor|virginica)$"

parse_species_tag <- function(value) {
  m <- regmatches(tolower(trimws(value)), regexec(SPECIES_TAG, tolower(trimws(value))))[[1]]
  if (length(m) == 0) {
    return(list(genus = NA_character_, name = value, matched = FALSE))
  }
  list(genus = if (nzchar(m[2])) m[2] else NA_character_, name = m[3], matched = TRUE)
}

#' Apply the regex across the frame.
tag_species <- function(df) {
  parsed <- lapply(df$species, parse_species_tag)
  df$species_matched <- vapply(parsed, function(p) p$matched, logical(1))
  log_step("tag_species", df)
}
